src/DssDirectDepositAaveDai.sol

```diff
@@ -246,94 +282,144 @@ contract DssDirectDepositAaveDai {
         emit Wind(amount);
     }
 
-    function unwind(uint256 amount, uint256 fees) external {
-        require(wards[msg.sender] == 1 || live == 0, "DssDirectDepositAaveDai/not-authorized");
+    function _unwind(uint256 supplyReduction, uint256 availableLiquidity, uint256 mode) internal {
+        bytes32 ilk_ = ilk;
+        address vow = chainlog.getAddress("MCD_VOW");
+        address end;
+        uint256 adaiBalance = adai.balanceOf(address(this));
+
+        uint256 daiDebt;
+        if (mode == 1) {
+            // Normal mode or module just caged (no culled)
+            // debt is obtained from CDP ink (or art which is the same)
+            (daiDebt,) = vat.urns(ilk_, address(this));
+        } else if (mode == 2) {
+            // Module shutdown and culled
+            // debt is obtained from free collateral owned by this contract
+            daiDebt = vat.gem(ilk_, address(this));
+        } else if (mode == 3) {
+            // MCD caged
+            // debt is obtained from free collateral owned by the End module
+            end = chainlog.getAddress("MCD_END");
+            EndLike(end).skim(ilk_, address(this));
+            daiDebt = vat.gem(ilk_, address(end));
+        } else if (mode == 4) {
+            // MCD caged but previously module was caged and culled
+            // debt is obtained from free collateral owned by the End module after some processing
+            daiDebt = vat.gem(ilk_, address(this));
+            end = chainlog.getAddress("MCD_END");
+            if (daiDebt > 0) {
+                require(daiDebt < 2 ** 255, "DssDirectDepositAaveDai/overflow");
+                vat.grab(ilk_, address(this), address(this), vow, int256(daiDebt), int256(daiDebt));
+                EndLike(end).skim(ilk_, address(this));
```

Comment:
I _think_ we could just `flux` the `gem` to the End (saves some gas)...unless there's some possibility that `tag` for this `ilk` will not be one. Doesn't hurt to do it the proper way, I suppose.

Reply:
Yes there is an easier way to do it with `slip`, but as you said doing via the proper methods looks more correct.
Regarding a different `tag`, this whole module assumes two things are true:
1) `rate` will be always 1
2) price, then `tag` will be always 1 (or maybe "at least 1")

For the second one, if it is lower than 1 we will start filling in `End.gap[ilk]`, which will provoke that we can't `slip` from the End anymore for unwinding as we will fuck up the whole accounting. So I think we better keep those two premises always true for this module. We should document them well.